### PR TITLE
741 army joinleave sync

### LINKED ISSUE #741: Army Join/Leave Sync
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

1. Sync the values in the **Synced** list below.
2. Create an integration test for all values (1 for server changing the value and 1 for client changing the value if applicable)
3. Create a command to test changing the value in game [see similar command](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/blob/development/source/GameInterface/Services/Towns/Commands/TownDebugCommand.cs#L455)
4. Do not allow players to join Armies that contain other players, to determine this use the `IsPlayerParty` extension method for MobileParty
5. Add a TODO comment to add player interactions where step 4 was implemented

Also see additional information at the bottom for more information and templates

**Key**
Server side only (only allow the running of the function on the server side)
Client side only (only allow the running of the function on the client side)
Client side sync (request the server to change the value, server then changes the value and replies to the client allowing the change)
Server side sync (only server allows running of the function and send to all clients the value changed and clients update the value/call original function)

**Synced**
Army.OnAddPartyInternal() => Server side sync
Army.OnRemovePartyInternal() => Server side sync
MobileParty.Army (Setter) => Server side sync

**Deferred**
None

**Non-Synced**
None

## Intended Design

**Message networking**
image
**Integration Testing**
image
## Location

Create a branch based from [development](https://github.com/Bannerlord-Coop-Team/BannerlordCoop)

## Related Issues

N/A

## Requirements

N/A

## Additional information

Using the [GameInterface Service Tempale](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/GameInterface/Services/Template) create commands to test the synced values.

For network communication use [Server Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Server/Services/Template) and [Client Service Template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.Core/Client/Services/Template)

It is recommended to start with an integration test, you can create one using the [test template](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/tree/development/source/Coop.IntegrationTests/Template).

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by thomas-soutif**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behaviour?

No sync when a mobile party join an army, or leave it

## What is the new behaviour?

Resolves #741

Sync when a party join an army, and a party leave an army

## Other information

### CODE REVIEW

File: source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs [hunk 1]
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
...
+    {
+
+        var stringBuilder = new StringBuilder();
+
+
+        if (args.Count != 1)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_list <ArmyId>");
+            return stringBuilder.ToString();
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs]:
```suggestion
            return "Usage: coop.debug.army.mobile_party_list <ArmyId>";
```

File: source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs [hunk 2]
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
...
+    {
+
+        var stringBuilder = new StringBuilder();
+
+
+        if (args.Count != 2)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_add <ArmyId> <MobilePartyId>");
+            return stringBuilder.ToString();
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs]:
```suggestion
            return "Usage: coop.debug.army.mobile_party_add <ArmyId> <MobilePartyId>";
```

File: source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs [hunk 3]
```diff
@@ -178,4 +181,138 @@ private static string GetArmyDispersionReasonUsage(StringBuilder stringBuilder =
...
+    {
+
+        var stringBuilder = new StringBuilder();
+
+
+        if (args.Count != 2)
+        {
+
+            stringBuilder.AppendLine("Usage: coop.debug.army.mobile_party_remove <ArmyId> <MobilePartyId>");
+            return stringBuilder.ToString();
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Commands/ArmyDebugCommand.cs]:
```suggestion
            return "Usage: coop.debug.army.mobile_party_remove <ArmyId> <MobilePartyId>";
```

File: source/GameInterface/Services/Armies/Patches/ArmyPatches.cs [hunk 4]
```diff
@@ -16,45 +20,87 @@ namespace GameInterface.Services.Armies.Patches;
...
+                /*foreach (var mobileParty_local in army.Parties)
+                {
+                    // Apply this code if duplication found
+                    if (mobileParty_local.StringId == mobileParty.StringId)
+                    {
+                        return;
+                    }
+                }*/
                 ArmyExtensions.AddPartyInternal(mobileParty, army);
+
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Patches/ArmyPatches.cs]: Remove dead code and use army extension
```suggestion
                army.AddPartyInternal(mobileParty);

```

File: source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs [hunk 5]
```diff
@@ -25,13 +25,13 @@ internal static class ArmyExtensions
             .GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static)
             .BuildDelegate<Action<Army, Army.ArmyDispersionReason>>();
 
-    internal static void AddPartyInternal(this MobileParty mobileParty, Army army)
+    internal static void AddPartyInternal(MobileParty mobileParty, Army army)
     {
         Army_OnAddPartyInternal(army,mobileParty);
     }
-    internal static void RemovePartyInternal(this MobileParty mobileParty, Army army)
+    internal static void RemovePartyInternal(MobileParty mobileParty, Army army)
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs]: make this into an army extension
```suggestion
    internal static void RemovePartyInternal(this Army army, MobileParty mobileParty)
```

File: source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs [hunk 6]
```diff
@@ -25,13 +25,13 @@ internal static class ArmyExtensions
             .GetMethod("ApplyInternal", BindingFlags.NonPublic | BindingFlags.Static)
             .BuildDelegate<Action<Army, Army.ArmyDispersionReason>>();
 
-    internal static void AddPartyInternal(this MobileParty mobileParty, Army army)
+    internal static void AddPartyInternal(MobileParty mobileParty, Army army)
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Extensions/ArmyExtensions.cs]:
```suggestion
    internal static void AddPartyInternal(this Army army, MobileParty mobileParty)
```

**Review by garrettluskey — CHANGES_REQUESTED** [review]

**Comment by thomas-soutif** [comment]
still have the bug when party add to army : duplication found
I've tried to send the List of Parties and setting it with the client every time a parties is add (to have the same list), it works almost fine, but there's a bug where sometimes, an army is not retrieve in the store (client side), so the mobile party cannot be add to the client army

This is the code I use to set directly a list (to give some idea) : 
```c#
internal static void AddPartyList(this Army army, List<MobileParty> mobilePartyListToUpdate)
{
    //Army_OnAddPartyInternal(army,mobileParty);
    
    // Get the _parties field using reflection
    FieldInfo partiesField = typeof(Army).GetField("_parties", BindingFlags.NonPublic | BindingFlags.Instance);

    
    MBList<MobileParty> newList = new MBList<MobileParty>();
    foreach (MobileParty party in mobilePartyListToUpdate)
    {
        newList.Add(party);
    }
    partiesField.SetValue(army, newList);
    
}
```

File: source/GameInterface/Services/Armies/Patches/ArmyPatches.cs [hunk 7]
```diff
@@ -16,45 +18,79 @@ namespace GameInterface.Services.Armies.Patches;
...
         if (PolicyProvider.AllowOriginalCalls) return true;
 
         if (ModInformation.IsClient) return false;
+        
 
-        var message = new MobilePartyInArmyRemoved(mobileParty.StringId, __instance.LeaderParty.StringId);
+        if (ContainerProvider.TryResolve<IObjectManager>(out var objectManager) == false)
+        {
+            Logger.Error("Unable to resolve {objectManager}", typeof(IObjectManager));
+            return false;
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Patches/ArmyPatches.cs]: If there is a failure this should return true to allow original method

File: source/GameInterface/Services/Armies/Patches/ArmyPatches.cs [hunk 8]
```diff
@@ -16,45 +18,79 @@ namespace GameInterface.Services.Armies.Patches;
...
+        }
+        
         
+        string armyId = __instance.GetStringId();
+        if (armyId == null)
+        {
+            Logger.Error("{army} was not properly registered", mobileParty.Army.Name);
+            return false;
```
Comment (garrettluskey) [inline source/GameInterface/Services/Armies/Patches/ArmyPatches.cs]: If there is a failure this should return true to allow original method

**Comment by thomas-soutif** [comment]
image

Has we can see, the main problem is that the client receive first mobile party add to army, before receiving the message to create the Army.
Here, the Army the Object Manager try to get is the CoopArmy_3
But in the game states, there's only 2 army the client have : The CoopArmy_1 and the CoopArmy_2

A solution could be to add a queue for mobile party add to army. Like this, if the army is not found because the client does not received it yet, it put it in a queue, and try later to add it (500ms or 1 second later)

**Comment by thomas-soutif** [comment]
I've implement a ArmyQueueManager to solved the issue I told before. Open for discussion, on optimization etc

**Comment by thomas-soutif** [comment]
Duplication fixed, and also the add / remove party sync

**Comment by thomas-soutif** [comment]
I've tested the new commits from @garrettluskey,
The sync of the army works well, the server is waiting for all clients to have the army created, but it creates also a big freeze every time an army is created (I think keeping the creation async with a queue system would be better for player experience)

The sync of mobile parties doesn't work well, a lot of duplication, parties are add to the army from nowhere( for the client).
@garrettluskey said it will be unstable until we fully sync all **party**, **hero** and **army data**. 

Finally, we cannot debug the game because of a crash that happened very quickly :
image
I was able to have 2 army created, but then a crash

I cleaned the project, and created a new party before test all of this.
So for now I don't know what we're doing with this branch, it's too unstable to be merge to the develop, and I'll not continue working on it if the rest is not implemented. So I propose to make a pause with this branch and start working on the other dependencys, I wait for the rest

**Review by garrettluskey — APPROVED** [review]

**Review by garrettluskey — APPROVED** [review]

**Pull request merged**